File: src/Keys.cs
using System;
using System.Security.Cryptography;
using System.Text;
using Org.BouncyCastle.Asn1.Sec;
using Org.BouncyCastle.Asn1.X9;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Security;

namespace TonMarket.src
{
    internal class Keys
    {
        public byte[] PublicKey;
        public byte[] PrivateKey;

        public string PublicKeyHex;
        public string PrivateKeyHex;
    }

    internal class KeysGenerator
    {
        public static int mnemonicLenght = 12;
        private static string mnemonicPathFile = "./wordlist/words.txt";

        public int privateKeyIterations = 10000;
        public int privateKeySize = 32;

        public string[] mnemonicPhrase { get; set; }


        private byte[] GeneratePrivateKey()
        {
            string concatenatedMnemonic = string.Join(" ", mnemonicPhrase);
            byte[] salt = Encoding.UTF8.GetBytes("mnemonic" + concatenatedMnemonic);

            using (var pbkdf2 = new Rfc2898DeriveBytes(concatenatedMnemonic, salt, privateKeyIterations, HashAlgorithmName.SHA256))
            {
                return pbkdf2.GetBytes(privateKeySize);
            }
        }

        private string ConvertToHex(byte[] byteArray)
        {
            StringBuilder hex = new StringBuilder(byteArray.Length * 2);
            foreach (byte b in byteArray)
            {
                hex.AppendFormat("{0:x2}", b);
            }
            return hex.ToString();
        }

        private byte[] GeneratePublicKey(byte[] privateKey)
        {
            // Use BouncyCastle to convert the private key and generate the public key
            X9ECParameters curve = SecNamedCurves.GetByName("secp256k1");
            ECDomainParameters domainParams = new ECDomainParameters(curve.Curve, curve.G, curve.N, curve.H);

            BigInteger d = new BigInteger(1, privateKey);
            ECPrivateKeyParameters privateKeyParams = new ECPrivateKeyParameters(d, domainParams);
            ECPublicKeyParameters publicKeyParams = new ECPublicKeyParameters(curve.G.Multiply(d), domainParams);

            return publicKeyParams.Q.GetEncoded();
        }

        public Keys GenerateKeys()
        {
            byte[] privateKey = GeneratePrivateKey();
            byte[] publicKey = GeneratePublicKey(privateKey);

            string privateKeyHex = ConvertToHex(privateKey);
            string publicKeyHex = ConvertToHex(publicKey);

            return new Keys()
            {
                PrivateKey = privateKey,
                PublicKey = publicKey,
                PrivateKeyHex = privateKeyHex,
                PublicKeyHex = publicKeyHex
            };
        }
    }
}

File: backend/src/keys/Keys.py
import os
import hashlib
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.backends import default_backend
from src.keys.Mnemonic import MnemonicGenerator

class Keys:

    def __init__(self, privateKeyHex, publicKeyHex):
        self.privatekey = privateKeyHex
        self.publickey = publicKeyHex


class KeysGenerator():

    def __init__(self, mnemonic_generator: MnemonicGenerator):
        self.mnemonic_generator = mnemonic_generator

    def GeneratePrivateKey(self):
        concatenatedMnemonic = " ".join(self.mnemonic_generator.getPhrase())
        salt = os.urandom(32)
        privatekey = hashlib.pbkdf2_hmac('sha256', concatenatedMnemonic.encode('utf-8'), salt, 100000, dklen=32)
        return privatekey

    def GeneratePublicKey(self, privateKey):
        curve = ec.SECP256K1()
        private_key_obj = ec.derive_private_key(int.from_bytes(privateKey, byteorder='big'), curve, default_backend())
        public_key_obj = private_key_obj.public_key()
        return public_key_obj.public_bytes(encoding=Encoding.X962, format=PublicFormat.UncompressedPoint)

    def ConvertToHex(self, byteArray):
        hex_string = ""
        for b in byteArray:
            hex_string += "{:02X}".format(b)
        return hex_string

    def GenerateKeys(self):
        privateKey = self.GeneratePrivateKey()
        publicKey = self.GeneratePublicKey(privateKey)

        privateKeyHex = self.ConvertToHex(privateKey)
        publicKeyHex = self.ConvertToHex(publicKey)

        keys = Keys(privateKeyHex, publicKeyHex)

        return keys
